### DISCUSSION TITLE
What's the difference between Lx = b and (I - M^-1 * L) x = b?

### DISCUSSION THREAD

**Discussion by jingma-git** [libigl/libigl#2210] [Q&A]
What's the difference between Lx = b and (I - M^-1 * L) x = b? 
L is laplacian matrix, M is mass matrix, I is identity matrix.

- Lx = b
    - Poisson equation, static, hard constraint?
-  (I - M^-1 * L) x = b? 
    - Diffusion, dynamic, time related, soft constraint?

**Comment by alecjacobson** [comment]
A couple things come to mind. L will generally be rank deficient (constant functions in the null space). So, L x = b will have many solutions.

I prefer to think of these equations as the Euler-Lagrange equations (or equilibrium equation) of energy minimization problems.

So, for example,

min_x ½∫ ‖ ∇x ‖² dA - ∫ x b dA 

models finding a low variation x under a load b.

For piecewise-linear functions, x. then 

∫ ‖ ∇x ‖² dA  = xᵀ L x 
∫ xb dA = xᵀ M b

so the discrete energy looks like

min_x ½ xᵀ L x - xᵀ M b

and the minimizer must satisfy:

L x = M b 

Meanwhile, we could consider a slightly different energy:

min_x ½∫ ‖ ∇x ‖² dA + ½∫ (x - b)² dA 

models finding a low variation x balanced against a data term pulling x toward b

∫ (x - b)² dA  = (x-b)ᵀ M (x-b)

So the discrete energy is

min_x ½ xᵀ L x + ½ (x-b)ᵀ M (x-b)

and the minimizer must satisfy

(L + M) x = M b

These equations are _close_ to what you have. Some things to be careful with:
 - the sign of L. Different papers/libraries use different signs. In my math above, I define L to be positive definite.
 - where the mass matrix M appears. The nice thing about starting with energies is that it's clear how M appears from integrals over the domain. Another way of double checking if M is appearing in the correct places is to do a units analysis.
- M⁻¹ L terms can almost always be avoided by multiplying through by M. The advantage to doing this is to preserve the sparsity of the system matrix being inverted.
